recipes_source/distributed_comm_debug_mode.rst

```diff
@@ -0,0 +1,197 @@
+Using CommDebugMode
+=====================================================
+
+**Author**: `Anshul Sinha <https://github.com/sinhaanshul>`__
+
+Prerequisites:
+
+- `Distributed Communication Package - torch.distributed <https://pytorch.org/docs/stable/distributed.html>`__
+- Python 3.8 - 3.11
+- PyTorch 2.2
+
+
+What is CommDebugMode and why is it useful
+------------------------------------------
+As the size of models continues to increase, users are seeking to leverage various combinations
+of parallel strategies to scale up distributed training. However, the lack of interoperability
+between existing solutions poses a significant challenge, primarily due to the absence of a
+unified abstraction that can bridge these different parallelism strategies. To address this
+issue, PyTorch has proposed DistributedTensor(DTensor) which abstracts away the complexities of
+tensor communication in distributed training, providing a seamless user experience. However,
+this abstraction creates a lack of transparency that can make it challenging for users to
+identify and resolve issues. To address this challenge, CommDebugMode, a Python context manager
+will serve as one of the primary debugging tools for DTensors, enabling users to view when and
+why collective operations are happening when using DTensors, addressing this problem.
+
+
+How to use CommDebugMode
+------------------------
+Using CommDebugMode and getting its output is very simple.
+
+.. code-block:: python
+
+    comm_mode = CommDebugMode()
+        with comm_mode:
+            output = model(inp)
+
+    # print the operation level collective tracing information
+    print(comm_mode.generate_comm_debug_tracing_table(noise_level=0))
+
+    # log the operation level collective tracing information to a file
+    comm_mode.log_comm_debug_tracing_table_to_file(
+        noise_level=1, file_name="transformer_operation_log.txt"
+    )
+
+    # dump the operation level collective tracing information to json file,
+    # used in the visual browser below
+    comm_mode.generate_json_dump(noise_level=2)
+
+    """
+    This is what the output looks like for a MLPModule at noise level 0
+    Expected Output:
+        Global
+          FORWARD PASS
+            *c10d_functional.all_reduce: 1
+            MLPModule
+              FORWARD PASS
+                *c10d_functional.all_reduce: 1
+                MLPModule.net1
+                MLPModule.relu
+                MLPModule.net2
+                  FORWARD PASS
+                    *c10d_functional.all_reduce: 1
+    """
```

Comment:
This sounds like a separate output to me not the part, you want users to copy paste?
```suggestion

 This is what the output looks like for a MLPModule at noise level 0:
 
 .. code-block:: bash
    Expected Output:
        Global
          FORWARD PASS
            *c10d_functional.all_reduce: 1
            MLPModule
              FORWARD PASS
                *c10d_functional.all_reduce: 1
                MLPModule.net1
                MLPModule.relu
                MLPModule.net2
                  FORWARD PASS
                    *c10d_functional.all_reduce: 1
```